File: src/revcomp/revcomp.nim
import std/strutils

const LineWidth = 60

proc complement(ch: char): char =
  case ch.toUpperAscii()
  of 'A': 'T'
  of 'C': 'G'
  of 'G': 'C'
  of 'T': 'A'
  of 'U': 'A'
  of 'M': 'K'
  of 'R': 'Y'
  of 'W': 'W'
  of 'S': 'S'
  of 'Y': 'R'
  of 'K': 'M'
  of 'V': 'B'
  of 'H': 'D'
  of 'D': 'H'
  of 'B': 'V'
  else: 'N'

proc flushRecord(header: string; sequence: string) =
  if header.len == 0:
    return
  stdout.write(header)
  stdout.write('\n')
  var line = newString(LineWidth)
  var filled = 0
  for i in countdown(sequence.len - 1, 0):
    line[filled] = complement(sequence[i])
    inc filled
    if filled == LineWidth:
      stdout.write(line)
      stdout.write('\n')
      filled = 0
  if filled > 0:
    stdout.write(line[0 ..< filled])
    stdout.write('\n')

when isMainModule:
  var header = ""
  var sequence = newStringOfCap(262144)
  for rawLine in stdin.readAll().splitLines():
    let line = rawLine.strip(chars = {'\r'})
    if line.len == 0:
      continue
    if line[0] == '>':
      flushRecord(header, sequence)
      header = line
      sequence.setLen(0)
    else:
      for ch in line:
        sequence.add(ch.toUpperAscii())
  flushRecord(header, sequence)

File: src/revcomp/revcomp.ml
let line_width = 60

let complement = function
  | 'A' | 'a' -> 'T'
  | 'C' | 'c' -> 'G'
  | 'G' | 'g' -> 'C'
  | 'T' | 't' -> 'A'
  | 'U' | 'u' -> 'A'
  | 'M' | 'm' -> 'K'
  | 'R' | 'r' -> 'Y'
  | 'W' | 'w' -> 'W'
  | 'S' | 's' -> 'S'
  | 'Y' | 'y' -> 'R'
  | 'K' | 'k' -> 'M'
  | 'V' | 'v' -> 'B'
  | 'H' | 'h' -> 'D'
  | 'D' | 'd' -> 'H'
  | 'B' | 'b' -> 'V'
  | _ -> 'N'

let flush_record header sequence =
  if header <> "" then begin
    print_endline header;
    let n = Buffer.length sequence in
    let text = Buffer.contents sequence in
    let line = Bytes.make line_width 'A' in
    let filled = ref 0 in
    for i = n - 1 downto 0 do
      Bytes.set line !filled (complement text.[i]);
      incr filled;
      if !filled = line_width then begin
        output stdout line 0 line_width;
        output_char stdout '\n';
        filled := 0
      end
    done;
    if !filled > 0 then begin
      output stdout line 0 !filled;
      output_char stdout '\n'
    end
  end

let () =
  let header = ref "" in
  let sequence = Buffer.create 262_144 in
  (try
     while true do
       let line = input_line stdin in
       if line <> "" then
         if line.[0] = '>' then begin
           flush_record !header sequence;
           header := line;
           Buffer.clear sequence
         end else
           String.iter (fun ch -> Buffer.add_char sequence (Char.uppercase_ascii ch)) line
     done
   with End_of_file -> ());
  flush_record !header sequence
